drivers/ble/sphero/sphero_r2d2_driver.go

```diff
@@ -0,0 +1,474 @@
+
+	r2AntiDosChara  = "00020005574f4f2053706865726f2121" // handshake 1st transmission
+	r2CommandsChara = "00010002574f4f2053706865726f2121" // send uuid
+	r2ResponseChara = r2CommandsChara
+
+	// command safe interval
+	commandInterval = time.Duration(12) * time.Millisecond
+
+	// start of packet
+	sop = 0x8D
+	// end of packet
+	eop        = 0xD8
+	escapeHex  = 0xAB
+	escapeMask = 0x88
+
+	// flags
+	isResponse                = 0b1
+	requestsResponse          = 0b10
+	requestsOnlyErrorResponse = 0b100
+	isActivity                = 0b1000
+	hasTargetId               = 0b10000
+	hasSourceId               = 0b100000
+	unused                    = 0b1000000
+	extendedFlags             = 0b10000000
+)
+
+// R2D2Driver is the Gobot driver for the Sphero R2D2 robot
+type R2D2Driver struct {
+	*ble.Driver
+	gobot.Eventer
+	defaultCollisionConfig spherocommon.CollisionConfig
+	seq                    uint8
+	seqMutex               sync.Mutex
+	collisionResponse      []uint8
+	packetChannel          chan *packet
+	asyncBuffer            []byte
+	asyncMessage           []byte
+	locatorCallback        func(p Point2D)
+	powerstateCallback     func(p spherocommon.PowerStatePacket)
+}
+
+// NewR2D2Driver creates a driver for a Sphero R2D2
+func NewR2D2Driver(a gobot.BLEConnector, opts ...ble.OptionApplier) *R2D2Driver {
+	return newR2D2BaseDriver(a, "R2D2", r2d2DefaultCollisionConfig(), opts...)
+}
+
+func newR2D2BaseDriver(
+	a gobot.BLEConnector, name string,
+	dcc spherocommon.CollisionConfig, opts ...ble.OptionApplier,
+) *R2D2Driver {
+	d := &R2D2Driver{
+		defaultCollisionConfig: dcc,
+		Eventer:                gobot.NewEventer(),
+		packetChannel:          make(chan *packet, 1024),
+		seqMutex:               sync.Mutex{},
+	}
+	d.Driver = ble.NewDriver(a, name, d.initialize, d.shutdown, opts...)
+
+	d.AddEvent(spherocommon.ErrorEvent)
+	d.AddEvent(spherocommon.CollisionEvent)
+
+	return d
+}
+
+// Wake wakes R2D2 up so we can play
+func (d *R2D2Driver) Wake() error {
+	// Power.wake did: 19 cid: 13
+	d.sendCraftPacket([]uint8{}, 0x13, 0x0D)
+
+	return nil
+}
+
+func (d *R2D2Driver) ResetLocatorData() {
+	// did: 24, cid: 19
+	d.sendCraftPacket([]uint8{}, 0x18, 0x13)
+}
+
+// ConfigureCollisionDetection configures the sensitivity of the detection.
+func (d *R2D2Driver) ConfigureCollisionDetection(cc spherocommon.CollisionConfig) {
+	// did: 24, cid: 17
+	d.sendCraftPacket([]uint8{cc.Method, cc.Xt, cc.Yt, cc.Xs, cc.Ys, cc.Dead}, 0x18, 0x11)
+}
+
+// GetLocatorData calls the passed function with the data from the locator
+func (d *R2D2Driver) GetLocatorData(f func(p Point2D)) {
+	// CID 0x15 is the code for the locator request
+	d.sendCraftPacket([]uint8{}, 0x02, 0x15)
+	d.locatorCallback = f
+}
+
+// GetPowerState calls the passed function with the Power State information from the sphero
+func (d *R2D2Driver) GetPowerState(f func(p spherocommon.PowerStatePacket)) {
+	// did: 19, cid: 4
+	d.sendCraftPacket([]uint8{}, 0x13, 0x04)
+	//	CHARGED = 0
```

Comment:
There are notes I copied from the spherov2 python lib. I might want to add them to the api docs as a short term fix or implement this correctly in a future commit.